chain/chain/src/chain.rs

```diff
@@ -2734,9 +2733,43 @@ impl<'a> ChainUpdate<'a> {
             Some(&block.hash()),
         )?;
         self.chain_store_update.save_block_extra(&block.hash(), BlockExtra { challenges_result });
+
         let protocol_version =
             self.runtime_adapter.get_epoch_protocol_version(block.header().epoch_id())?;
 
+        // This part of code re-introduces outgoing receipts lost because of a bug in apply_chunks
+        // (see https://github.com/near/nearcore/pull/4248/)
+        // We take the first block in the first epoch in which protocol feature RestoreReceiptsAfterFix
+        // is enabled, and save the restored receipts there.
+        if self.runtime_adapter.is_next_block_epoch_start(prev_block.hash()).unwrap_or(false) {
+            checked_feature!(
+                "protocol_feature_restore_receipts_after_fix",
+                RestoreReceiptsAfterFix,
+                protocol_version,
+                {
+                    let prev_protocol_version = self
+                        .runtime_adapter
+                        .get_epoch_protocol_version(prev_block.header().epoch_id())?;
+                    if !checked_feature!(
+                        "protocol_feature_restore_receipts_after_fix",
+                        RestoreReceiptsAfterFix,
+                        prev_protocol_version
+                    ) {
+                        let receipt_result = self
+                            .runtime_adapter
+                            .get_protocol_config(block.header().epoch_id())?
+                            .runtime_config
+                            .receipts_to_restore;
+                        self.chain_store_update.save_outgoing_receipt(
+                            &block.hash(),
+                            0,
+                            receipt_result,
+                        );
+                    }
+                }
+            );
+        }
```

Comment:
I am not sure why you want to save those as outgoing receipts. Even if you decide to do that, the implementation is not correct -- you need to change the outgoing receipt roots as well. Otherwise the invariant that outgoing receipts for a given block and shard match its outgoing receipt root in the corresponding chunk would be broken.

Reply:
Agree, moved receipts addition to `process_state_update`.